File: ios/Sources/KlinikSurveysFeature/SurveyModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum SurveyPhase: Sendable, Equatable {
    case loading
    case loaded
    /// Nothing to fill in. Not a failure.
    case none
    case failed(String)
}

public struct SurveyState: Sendable, Equatable {
    public var phase: SurveyPhase = .loading
    public var surveys: [PendingSurvey] = []
    /// Answers for the survey on screen, keyed by question.
    public var answers: [String: SurveyAnswer] = [:]
    public var submitting = false
    public var submitted = false
    /// The answers were kept on the phone rather than delivered (spec M15).
    public var queued = false
    public var error: String?

    public init() {}

    /// The one being answered: the earliest still open.
    public var current: PendingSurvey? {
        surveys.first { $0.isOpen() }
    }

    /// Whether every required question has an answer.
    public var canSubmit: Bool {
        guard let current else { return false }

        return current.questions
            .filter(\.isRequired)
            .allSatisfy { answers[$0.id] != nil }
    }
}

/**
 * The short surveys a patient fills in after surgery (spec M18).
 *
 * Answers are held until the whole form is sent. A survey saved question by
 * question would leave half-answered forms in the record that read as a patient
 * reporting nothing about the other half — and a pain score of "nothing" is a
 * clinical statement, not an absence.
 *
 * A closed survey is shown as closed rather than hidden. Somebody who opens the
 * app a week after the reminder should be told they missed it, not left
 * wondering whether the app forgot.
 */
@MainActor
public final class SurveyModel {
    private let api: SurveysAPI
    private var state = SurveyState()

    public init(api: SurveysAPI) {
        self.api = api
    }

    public func currentState() -> SurveyState { state }

    public func load() async {
        do {
            state.surveys = try await api.mine()
                .sorted { $0.scheduledFor < $1.scheduledFor }
            state.phase = state.surveys.isEmpty ? .none : .loaded
            state.answers = [:]
            state.submitted = false
        } catch let error as APIError {
            state.phase = .failed(L10n.message(for: error))
        } catch {
            state.phase = .failed(L10n.string("error.server"))
        }
    }

    public func answer(_ questionId: String, _ answer: SurveyAnswer) {
        state.answers[questionId] = answer
    }

    public func submit() async {
        guard let survey = state.current, state.canSubmit else { return }

        state.submitting = true
        state.error = nil

        defer { state.submitting = false }

        do {
            _ = try await api.submit(survey.id, answers: state.answers)

            state.surveys.removeAll { $0.id == survey.id }
            state.answers = [:]
            state.submitted = true
            state.queued = false
            state.phase = state.surveys.contains { $0.isOpen() } ? .loaded : .none
        } catch APIError.queuedForLater {
            // Ten minutes of somebody's attention, kept. The questionnaire
            // leaves the list exactly as it would have: it is answered, and
            // what is outstanding is the delivery, not the answering.
            state.surveys.removeAll { $0.id == survey.id }
            state.answers = [:]
            state.submitted = true
            state.queued = true
            state.error = nil
            state.phase = state.surveys.contains { $0.isOpen() } ? .loaded : .none
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }
    }
}

// MARK: - Staff side

public enum SurveyTrendPhase: Sendable, Equatable {
    case loading
    case loaded(PatientSurveys)
    case empty
    case failed(String)
}

/// One patient's answers over time, for the clinician (spec M18).
@MainActor
public final class SurveyTrendModel {
    private let api: SurveysAPI
    private let patientId: String

    private var phase: SurveyTrendPhase = .loading

    public init(api: SurveysAPI, patientId: String) {
        self.api = api
        self.patientId = patientId
    }

    public func currentPhase() -> SurveyTrendPhase { phase }

    public func load() async {
        do {
            let surveys = try await api.forPatient(patientId)
            phase = surveys.series.isEmpty ? .empty : .loaded(surveys)
        } catch let error as APIError {
            phase = .failed(L10n.message(for: error))
        } catch {
            phase = .failed(L10n.string("error.server"))
        }
    }
}

File: android/feature/surveys/src/main/kotlin/xyz/klinik/feature/surveys/SurveyModel.kt
package xyz.klinik.feature.surveys

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.ApiError
import xyz.klinik.network.PendingSurvey
import xyz.klinik.network.SurveyAnswer
import xyz.klinik.network.SurveysApi
import xyz.klinik.network.UiText
import xyz.klinik.network.uiText

sealed interface SurveyPhase {
    data object Loading : SurveyPhase
    data object Loaded : SurveyPhase

    /** Nothing to fill in. Not a failure. */
    data object None : SurveyPhase
    data class Failed(val message: UiText) : SurveyPhase
}

data class SurveyState(
    val phase: SurveyPhase = SurveyPhase.Loading,
    val surveys: List<PendingSurvey> = emptyList(),
    /** Answers for the survey on screen, keyed by question. */
    val answers: Map<String, SurveyAnswer> = emptyMap(),
    val submitting: Boolean = false,
    val submitted: Boolean = false,
    val error: UiText? = null,
    /** The clock the form is measured against, so "expired" is testable. */
    internal val now: String = "",
) {
    /** The one being answered: the earliest still open. */
    val current: PendingSurvey? get() = surveys.firstOrNull { it.isOpen(now) }

    /**
     * The ones whose window has closed.
     *
     * Shown as closed rather than hidden: somebody opening the app a week
     * after the reminder should be told they missed it, not left wondering
     * whether the app forgot to ask.
     */
    val missed: List<PendingSurvey> get() = surveys.filterNot { it.isOpen(now) }

    /** Whether every required question has an answer. */
    val canSubmit: Boolean
        get() = current?.requiredQuestions?.all { answers[it.id] != null } == true
}

/**
 * The short questionnaires a patient fills in after surgery (spec M18).
 *
 * Answers are held until the whole form is sent. A survey saved question by
 * question would leave half-answered forms in the record that read as a
 * patient reporting nothing about the rest — and a pain score of "nothing" is
 * a clinical statement, not an absence.
 *
 * Nothing here reads the answers back to the patient. The server computes
 * findings for the clinic; "your reported pain has worsened" is a clinical
 * reading, and a questionnaire screen is not the thing that should deliver
 * one.
 */
class SurveyModel(
    private val api: SurveysApi,
    /** ISO-8601 UTC, injected so an expiring form can be tested without waiting. */
    private val now: () -> String,
) {
    private val _state = MutableStateFlow(SurveyState())
    val state: StateFlow<SurveyState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = _state.value.copy(phase = SurveyPhase.Loading)

        try {
            val surveys = api.mine().sortedBy { it.scheduledFor }
            val moment = now()

            _state.value = SurveyState(
                phase = if (surveys.none { it.isOpen(moment) }) {
                    SurveyPhase.None
                } else {
                    SurveyPhase.Loaded
                },
                surveys = surveys,
                now = moment,
            )
        } catch (error: Throwable) {
            _state.value = _state.value.copy(phase = SurveyPhase.Failed(messageFor(error)))
        }
    }

    fun answer(questionId: String, answer: SurveyAnswer) {
        _state.value = _state.value.copy(answers = _state.value.answers + (questionId to answer))
    }

    suspend fun submit() {
        val survey = _state.value.current ?: return
        if (!_state.value.canSubmit) return

        _state.value = _state.value.copy(submitting = true, error = null)

        try {
            api.submit(survey.id, _state.value.answers)

            val remaining = _state.value.surveys.filterNot { it.id == survey.id }
            val moment = now()

            _state.value = _state.value.copy(
                phase = if (remaining.none { it.isOpen(moment) }) {
                    SurveyPhase.None
                } else {
                    SurveyPhase.Loaded
                },
                surveys = remaining,
                answers = emptyMap(),
                submitting = false,
                submitted = true,
                now = moment,
            )
        } catch (error: Throwable) {
            // The answers stay in `answers`. Ten minutes of somebody's
            // attention is not something to throw away because the network
            // dropped: the form is still on screen, filled in, with the reason
            // it did not go.
            _state.value = _state.value.copy(submitting = false, error = messageFor(error))
        }
    }

    private fun messageFor(error: Throwable): UiText =
        (error as? ApiError)?.uiText() ?: UiText.Key("error.server")
}
